source/fuzz/fuzzer_pass_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,157 @@
+
+void FuzzerPassChangingMemorySemantics::Apply() {
+  ForEachInstructionWithInstructionDescriptor(
+      [this](opt::Function* /* unused */, opt::BasicBlock* /* unused */,
+             opt::BasicBlock::iterator inst_it,
+             const protobufs::InstructionDescriptor& instruction_descriptor) {
+        // Instruction must be atomic or barrier instruction, return the number
+        // of memory semantic operands for instruction.
+        auto number_of_memory_semantics =
+            TransformationChangingMemorySemantics::GetNumberOfMemorySemantics(
+                inst_it->opcode());
+
+        // Equal zero if no atomic or barrier instruction is found.
+        if (number_of_memory_semantics == 0) {
+          return;
+        }
+
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfChangingMemorySemantics())) {
+          return;
+        }
+
+        // If the instruction has two memory semantic operands pick one
+        // randomly.
+        uint32_t memory_semantics_operand_position = 0;
+        if (number_of_memory_semantics == 2) {
+          std::vector<uint32_t> operand_positions{0, 1};
+          memory_semantics_operand_position =
+              GetFuzzerContext()->RandomIndex(operand_positions);
+        }
+
+        auto needed_index = TransformationChangingMemorySemantics::
+            GetMemorySemanticsOperandIndex(inst_it->opcode(),
+                                           memory_semantics_operand_position);
+
+        auto memory_semantics_value =
+            GetIRContext()
+                ->get_def_use_mgr()
+                ->GetDef(inst_it->GetSingleWordInOperand(needed_index))
+                ->GetSingleWordInOperand(0);
+
+        auto lower_bits_old_memory_semantics =
+            static_cast<SpvMemorySemanticsMask>(
+                memory_semantics_value & TransformationChangingMemorySemantics::
+                                             kMemorySemanticsLowerBitmask);
+        auto higher_bits_old_memory_semantics =
+            static_cast<SpvMemorySemanticsMask>(
+                memory_semantics_value & TransformationChangingMemorySemantics::
+                                             kMemorySemanticsHigherBitmask);
+
+        std::vector<SpvMemorySemanticsMask> memory_semanitcs_masks{
+            SpvMemorySemanticsMaskNone, SpvMemorySemanticsAcquireMask,
+            SpvMemorySemanticsReleaseMask, SpvMemorySemanticsAcquireReleaseMask,
+            SpvMemorySemanticsSequentiallyConsistentMask};
+
+        // Remove the memory mask is not applicable for instruction.
+        memory_semanitcs_masks.erase(
+            std::remove_if(memory_semanitcs_masks.begin(),
+                           memory_semanitcs_masks.end(),
+                           [inst_it, lower_bits_old_memory_semantics](
+                               SpvMemorySemanticsMask /*unused*/) {
+                             switch (inst_it->opcode()) {
```

Comment:
I can't use this function as I talk with you because the `IsValidConversion` with its new name `IsSuitableStrengthening`, takes these arguments `function prototype`
```cpp
  static bool IsSuitableStrengthening(
      opt::IRContext* ir_context,
      spvtools::opt::Instruction* needed_atomic_instruction,
      SpvMemorySemanticsMask first_5bits_old_memory_semantics,
      SpvMemorySemanticsMask first_5bits_new_memory_semantics,
      uint32_t memory_semantics_operand_position, SpvMemoryModel memory_model);
```
it needs `first_5bits_new_memory_semantics` and I'm not generating it at this step.
What do you think @paulthomson?

Reply:
I can't use this function as I talk with you because the `IsValidConversion` with its new name `IsSuitableStrengthening`, takes these arguments:
```cpp
  // function prototype
  static bool IsSuitableStrengthening(
      opt::IRContext* ir_context,
      spvtools::opt::Instruction* needed_atomic_instruction,
      SpvMemorySemanticsMask first_5bits_old_memory_semantics,
      SpvMemorySemanticsMask first_5bits_new_memory_semantics,
      uint32_t memory_semantics_operand_position, SpvMemoryModel memory_model);
```
It needs `first_5bits_new_memory_semantics` and I'm not generating it at this step.
What do you think @paulthomson?

Reply:
Let's discuss this via video call. In short: `potential_new_memory_orders` is your list of first_5bits_new_memory_semantics. Also, while we are here, you should rename first_5bits to lower_bits, as per a previous comment of mine.

Reply:
Done.